### ISSUE TITLE
[RFC] Add support for distributed full-graph training and inference

### ISSUE THREAD

**Issue by hesham-mostafa** [pytorch/pytorch#91258]
### 🚀 The feature, motivation and pitch

This issue is for discussing the integration of the Sequential Aggregation and Rematerialization technique for distributed full-graph training and inference from [here](https://github.com/IntelLabs/SAR) into DGL. For large graphs, DGL supports distributed sampling-based training (DistDGL) where the graph is partitioned across several machines and distributed sampling takes care of creating the mini-batches for the trainers. There is no mechanism, however, to do either distributed full-graph training or distributed full-graph inference. Full-graph inference would be useful to get test/validation accuracies after doing sampling-based training. While you could do GNN inference using graph samples, the results would have some randomness. Taking the average predictions of several inference runs is possible but costly. Moreover, there is no guarantee that you are not discarding some crucial structural or node features.

There are many graph sampling methods out there, and some are problem specific (in fraud detection models for example), which would require new users to experiment to find the most performant method. Choosing the best sampling hyperparameters (such as neighbor sampling fanout) is also not trivial. Conceptually and in terms of hyper-parameter counts, full-graph training is much simpler. It would be beneficial to be able to train on the full graph (no sampling involved) as a sanity check and also to obtain a useful baseline to compare against the performance of various sampling-based training methods. 

The proposed implementation uses the graph partitions creates by DGL's metis (``dgl.distributed.partition_graph``) to construct a distributed graph object at each worker. This distributed graph object can be used as a drop-in replacement to ``dgl.DGLHeterograph`` and passed to a GNN. Under the hood, we take care of communicating the features of nodes at the partition boundaries in the forward pass, and collecting their gradients in the backward pass to realize a full-graph forward and backward GNN pass. We guarantee that no worker would need to materialize more than 2 graph partitions at once (even for densely connected graphs), and therefore obtain linear memory scaling, i.e, per-worker node/edge feature memory goes down linearly with number of workers.

### Additional context

For distributed full-graph inference, we can distribute the graph across several machines. However, when doing the message passing in a GNN layer, a machine might need to fetch a substantial part of the graph to obtain the neighborhood of its local nodes as shown below. This can exhaust the machine's memory. This problem is more severe in more densely connected graphs. 
![image](https://user-images.githubusercontent.com/16276103/208975616-69e2dfe0-aab5-4098-bfb8-fa0b29f816ef.png)

The simple solution depends on the fact that the majority of GNN reduction methods such as ``sum``, ``mean`` and ``max`` can be recursively applied, i.e, ``reduce(A,B,C) == reduce(A,reduce(B,C))``, which would allow us to do the message passing sequentially. Even functions as ``edge_softmax`` used in attention-based GNNs can be calculated sequentially. The sequential  message passing operation is shown below:

![image](https://user-images.githubusercontent.com/16276103/208975698-0b99dd53-43b5-4782-8d82-b4791fe0b524.png)

After fetching the neighboring nodes from only one remote worker and aggregating their messages, machine 1 can delete these remote nodes and go on to fetch the remote node features from the next worker. We thus obtain linear memory scaling in the full-graph inference pass, as each machine would at most need to materialize two graph partitions, its own partition and the remote partition whose messages are being actively aggregated. 

The situation becomes more complicated when doing full-graph training, as even if we fetch and aggregate messages from remote partitions one by one, the remotely fetched node features would live on in the local computational graph constructed by Autograd which would remove the memory benefits of sequential aggregation. To avoid that, we disable autograd when doing the distributed message passing step, i.e, the message passing calls (``apply_edges`` and ``update_all``) disable Autograd . We construct this computational graph piece by piece during the backward pass, where each piece involves only the data of two partitions as shown below:

![image](https://user-images.githubusercontent.com/16276103/208975874-59bd91a4-615d-42d8-bc73-b55b339cbfd0.png)

As machine 1 receives errors for its local nodes, it follows the following steps:
1. Backpropagate the errors in the local partition.
2. Construct the message passing computational graph involving only machine 1 and machine 2, backpropagtes the errors to machine 2, then delete the computational graph.
3. Construct the message passing computational graph involving only machine 1 and machine 3, backpropagtes the errors to machine 3, then delete the computational graph. 

Machines 2 and 3 follow similar steps. constructing the computational graph piece may involve re-fetching the remote features and running a partial message passing round. For many GNN layer types however, such as GraphSage, GCNII, etc.. we can backpropagate the gradients to remote machines without knowing the remote input features. That is because backpropagating errors through a sum or mean operator does not require that we know the operator's inputs. Our method detects such cases and backpropagates the errors directly without remote re-fetching.

In summary, we obtain linear memory scaling in both the forward and the backward pass by NOT constructing the prohibitively expensive message passing computational graph in the forward pass, but instead constructing it piece by piece during the backward pass. The computational graph for all operations other than the message passing operations is constructed normally during the forward pass.

## API and usage scenarios

We will re-use the same graph partitioning data generated by DistDGL's metis-based partitioning (``dgl.distributed.partition_graph``) to make it easy to switch between sampling-based and full-graph training:

![image](https://user-images.githubusercontent.com/16276103/208976091-0cbe6f5b-5aa0-4cfe-a342-279ba779ea7b.png)

We propose to add a sub-package to DGL: ``sar`` that can be used to run distributed full-graph training and inference as follows:
 
```python

#sar uses torch.distributed for all communication. Backend
#has to support all_to_all collective. 'nccl' backend for GPU training and 'ccl'
#backend for CPU training are recommended
torch.distributed.init_process_group('ccl',rank = rank,world_size = world_size)

#Load the partition data created by dgl.distributed.partition_graph
partition_data = dgl.sar.load_partition_data('path to json file',rank)

full_dist_graph = dgl.sar.construct_graph(partition_data) 

model = make_gnn_model()
model = torch.nn.parallel.DistributedDataParallel(model)

optimizer = torch.optim.Adam(model.parameters(),lr)

model.train()
for train_iter in range(n_train_iters):
    y = model(full_dist_graph,partition_data.node_features['feats'])
    loss = calculate_loss(y,partition_data.node_features['labels'],
			partition_data.node_features['train_mask'])
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    
model.eval()
with torch.no_grad():
     y = model(full_dist_graph,partition_data.node_features['feats'])
     valid_loss = calculate_loss(y,partition_data.node_features['labels'],
			partition_data.node_features['valid_mask'])
```

``full_dist_graph`` exposes a subset of the API of ``dgl.DGLHeteroGraph`` that is relevant for GNN training. In particular it implements a distributed memory-efficient version of ``update_all`` and ``apply_edges``. It can thus be passed directly to a standard DGL GNN.

### distdgl and sar

This diagram highlights the main differences between ``sar`` and ``distdgl``

![sar_rfc](https://user-images.githubusercontent.com/16276103/208977025-8d69eb06-973c-4c5f-a290-bac2510a1ad9.png)

``sar`` needs to run message passing on the distributed graph so it requires the graph to be compacted, i.e, node indices start from 0 and have no gaps. To enable sequential aggregation, we also need the graph partition at each machine to be split into multiple parts where each part includes the edges from only one remote partition. Finally, full-graph training and inference requires several synchronization points in the forward and backward passes. Our experiments indicate that synchronous all_to_all collectives are faster than point-to-point communication through Tensorpipe.

One usage scenario that we want to make as painless as possible is distributed sampling-based training using ``distdgl``followed by distributed full-graph inference using ``sar``. The simplest way to do that is:

```python

import dgl
import torch as torch

dgl.distributed.initialize('ip_config.txt')
torch.distributed.init_process_group(backend='ccl',rank = rank,world_size = world_size)
g = dgl.distributed.DistGraph('graph_name', 'part_config.json')

# .. rest of distdgl training pipeline ...

##Distributed inference using full graph
partition_data = dgl.sar.load_partition_data('part_config.json',rank)
full_dist_graph = dgl.sar.construct_graph(partition_data) 
model.eval()
with torch.no_grad():
     y = model(full_dist_graph,partition_data.node_features['feats'])
     valid_loss = calculate_loss(y,partition_data.node_features['labels'],
			partition_data.node_features['valid_mask'])

```

Alternatively, we can support direct conversion from ``DistGraph`` to ``GraphShardManager`` to avoid loading the partition data from disk twice.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by mrshenli** [comment]
Hey @hesham-mostafa, is the proposal to add a new package to DGL? In that case, I assume this RFC issue needs to be posted to DGL repo? https://github.com/dmlc/dgl

**Comment by hesham-mostafa** [comment]
Oops. My apologies. Wrong repo :) Will close this now.
